### ISSUE TITLE
 The data processing speed suddenly decreased

### ISSUE THREAD

**Issue by ForeverNightmare** [pkargupta/MEGClass#2]
Hi,
    I was training my dataset based on your model. For the class-oriented document representations generation step, at first the program ran very well like this
SENTENCE INFO! # of docs: 261465, # of sentences: 3826706
Starting to compute class-oriented document representations...
Finish reading data
.....
.....
.....
  0%|          | 0/3826706 [00:00<?, ?it/s]
  0%|          | 1/3826706 [00:08<9249:31:04,  8.70s/it]
  0%|          | 6/3826706 [00:08<1155:24:37,  1.09s/it]
  0%|          | 13/3826706 [00:08<429:14:43,  2.48it/s]
  0%|          | 20/3826706 [00:09<232:22:38,  4.57it/s]
  0%|          | 28/3826706 [00:09<136:32:58,  7.78it/s]
  0%|          | 36/3826706 [00:09<89:20:48, 11.90it/s] 
  0%|          | 44/3826706 [00:09<62:38:57, 16.97it/s]
  0%|          | 52/3826706 [00:09<46:30:47, 22.85it/s]
  0%|          | 60/3826706 [00:09<36:01:57, 29.50it/s]
  0%|          | 68/3826706 [00:09<28:52:52, 36.80it/s]
  0%|          | 76/3826706 [00:09<24:18:19, 43.73it/s]
  0%|          | 85/3826706 [00:09<20:31:35, 51.78it/s]
  0%|          | 93/3826706 [00:09<18:55:17, 56.18it/s]
  0%|          | 102/3826706 [00:10<17:02:58, 62.34it/s]
  0%|          | 110/3826706 [00:10<16:11:51, 65.62it/s]
  0%|          | 118/3826706 [00:10<15:40:41, 67.80it/s]

But after some times, its processing speed suddenly decreased:
 50%|█████     | 1926036/3826706 [8:06:20<53:25:23,  9.88it/s]
 50%|█████     | 1926036/3826706 [8:06:37<53:25:23,  9.88it/s]
 50%|█████     | 1926040/3826706 [8:06:37<507:14:58,  1.04it/s]
 50%|█████     | 1926041/3826706 [8:06:40<559:09:08,  1.06s/it]
 50%|█████     | 1926041/3826706 [8:06:57<559:09:08,  1.06s/it]
 50%|█████     | 1926043/3826706 [8:07:05<1573:05:14,  2.98s/it]
 50%|█████     | 1926044/3826706 [8:07:27<2591:48:12,  4.91s/it]
 50%|█████     | 1926045/3826706 [8:07:52<3992:03:08,  7.56s/it]
 50%|█████     | 1926046/3826706 [8:07:57<3724:24:52,  7.05s/it]
 50%|█████     | 1926048/3826706 [8:08:03<3047:30:30,  5.77s/it]
 50%|█████     | 1926048/3826706 [8:08:17<3047:30:30,  5.77s/it]
 50%|█████     | 1926050/3826706 [8:08:26<4030:37:08,  7.63s/it]
 50%|█████     | 1926051/3826706 [8:08:44<5002:18:24,  9.47s/it]
 50%|█████     | 1926052/3826706 [8:09:07<6454:10:31, 12.22s/it]
 50%|█████     | 1926053/3826706 [8:09:15<5914:50:08, 11.20s/it]
 50%|█████     | 1926054/3826706 [8:09:16<4559:24:37,  8.64s/it]
 50%|█████     | 1926055/3826706 [8:09:24<4554:02:34,  8.63s/it]
 50%|█████     | 1926056/3826706 [8:09:50<6964:51:48, 13.19s/it]
 50%|█████     | 1926057/3826706 [8:10:13<8388:51:31, 15.89s/it]
 50%|█████     | 1926058/3826706 [8:10:33<9045:47:35, 17.13s/it]
 50%|█████     | 1926059/3826706 [8:10:54<9621:06:17, 18.22s/it]
 50%|█████     | 1926060/3826706 [8:11:18<10487:34:24, 19.86s/it]

Have you met similar issue in your research?

**Comment by pkargupta** [comment]
Hm, that is quite odd, I didn't run into this issue on my end, but this may be due to the size of the datasets we use being significantly smaller than yours. It seems like the bottleneck is at [this line](https://github.com/pkargupta/MEGClass/blob/3066b51ad03784d10755effad7ff1345248f5cec/class_oriented_sent_representations.py#L286), which may be indicating that after a certain point, appending so many sentence representations to the `sent_representations` array may be consuming a lot of memory and becoming increasingly inefficient? We use a single NVIDIA RTX A6000 GPU for our largest dataset that uses approximately 33K documents for reference. I suggest that if increasing compute resources does not help or is not an option, you can save the class words that you have already learned from the entire dataset (save the `class_representations` array on line 233) since that is much more quick, and then split your dataset of documents. The class-oriented sentence representations are independently computed from each other if given the class representations, so the bottleneck on line 286 can be parallelizable. Hopefully this change will help; this portion of the code is unfortunately the most time-consuming out of all modules, so once the class-oriented sentence representations are computed in parallel, the main MEGClass framework should not be as inefficient. Please let me know if you have any other questions and if this solves the issue. Thanks!

**Comment by ForeverNightmare** [comment]
@pkargupta Thanks for your reply! I reduced the data sized to about 35mb and fished the class-oriented work successfully. But when running the  megclass.main(args), the program stopped for no reason after finishing iter 1 training. The iter 0 works well, all the training and evaluate output is shown in console. 
![image](https://github.com/pkargupta/MEGClass/assets/37356781/792fd157-d0a5-4aee-8a7e-450e066ab0b9)
![image](https://github.com/pkargupta/MEGClass/assets/37356781/64dc8a6c-5b24-4245-9a84-75b0b671bc9c)


Is that also because no more memory can be allocated?

**Comment by ForeverNightmare** [comment]
And, the “product” of megclass.main(args),{args.dataset_name}_model_e{args.epochs}.pth is not used in the next step, train_soft/text_classifier.main(args), but still has training and evaluation result. Is that correct? If so, which one part can we get the megclass's training and evaluation result from, megclass.main or train_soft/text_classifier.main?

**Comment by pkargupta** [comment]
> @pkargupta Thanks for your reply! I reduced the data sized to about 35mb and fished the class-oriented work successfully. But when running the megclass.main(args), the program stopped for no reason after finishing iter 1 training. The iter 0 works well, all the training and evaluate output is shown in console. ![image](https://user-images.githubusercontent.com/37356781/240432378-792fd157-d0a5-4aee-8a7e-450e066ab0b9.png) ![image](https://user-images.githubusercontent.com/37356781/240432926-64dc8a6c-5b24-4245-9a84-75b0b671bc9c.png)
> 
> Is that also because no more memory can be allocated?

Hm that is odd, I am running it on my end with this version of the code, and I do not seem to be running into this issue. Are you running it on the class-oriented sentence representations from your 261K document dataset or just the 35MB one? It may be a memory inefficiency issue again, but it is odd since only `class_set` is the only variable that could be affected significantly between the first and second iterations (adding in the confident documents to the class-indicative word-based class set), but it seems like your run is progressing to [line 44](https://github.com/pkargupta/MEGClass/blob/3066b51ad03784d10755effad7ff1345248f5cec/megclass.py#L44) in the second iteration. Do you know which line within the nested loop your run is getting stuck on? 
As a sanity check, maybe try using the reduced dataset or reduce parameter `--k` since the default percent of confident documents added to the class set is 7.5%, which can be quite large for 261K documents (19.6K docs).


> And, the “product” of megclass.main(args),{args.dataset_name}_model_e{args.epochs}.pth is not used in the next step, train_soft/text_classifier.main(args), but still has training and evaluation result. Is that correct? If so, which one part can we get the megclass's training and evaluation result from, megclass.main or train_soft/text_classifier.main?

The .pth file just stores the trained multi-head self attention model that has learned the contextualized sentence and document representations. megclass.main is essentially constructing a robust and confident pseudo dataset (saved on [line 167](https://github.com/pkargupta/MEGClass/blob/3066b51ad03784d10755effad7ff1345248f5cec/megclass.py#L167)) that is used to learn a final text classifier in train_text_classifier.py (or train_soft_classifier based on your preference according to your dataset). Hence, we can get the training and evaluation result by running train_text_classifier.py (this is already called for you in line 29 of run.py).


> Another issue , when I run the run .py's train_text_classifier seperately(by # previous lines in main() ), this error occured:
> 
> Training classifier with hard labels! Traceback (most recent call last): File "run4.py", line 241, in main(args) File "run4.py", line 29, in main train_text_classifier.main(args) File "/home/xuanw/HL/MEGClass-main/train_text_classifier.py", line 550, in main train_dataset = load_and_cache_examples(args, args.dataset_name, tokenizer, evaluate=False) File "/home/xuanw/HL/MEGClass-main/train_text_classifier.py", line 431, in load_and_cache_examples processor.get_test_examples(args.test_data_dir, args.dataset_name) if evaluate else processor.get_train_examples(args.train_data_dir) File "/home/xuanw/HL/MEGClass-main/train_text_classifier.py", line 58, in get_train_examples for i, (text, label) in enumerate(zip(load_text(os.path.join(data_dir, self.train_dir)), File "/home/xuanw/HL/MEGClass-main/preprocessing_utils.py", line 64, in load_text with open(os.path.join(data_dir, 'dataset.txt'), mode='r', encoding='utf-8') as text_file: FileNotFoundError: [Errno 2] No such file or directory: '/home/xuanw/HL/MEGClass-main/intermediate_data/ssmedic1112/dataset.txt'
> 
> It seems that train_text_classifier tried to load dataset.txt from intermediate data folder. But in the previous steps, there are only some .pk and .pth files were stored there.

The pseudo-dataset (chosen documents and their corresponding labels) that I mentioned above will be stored in the intermediate data folder as dataset.txt and labels.txt. In other words, train_text_classifier will only run if [line 167 in megclass.py](https://github.com/pkargupta/MEGClass/blob/3066b51ad03784d10755effad7ff1345248f5cec/megclass.py#L167) has run successfully. If you would like to see the results from the first iteration in the meantime while resolving the stuck second iteration, I would comment out the static + class_oriented representation lines as you mentioned, but pass in 1 as a value for the --iters argument. That way, the pseudo-dataset will get saved.

**Comment by ForeverNightmare** [comment]
@pkargupta Thanks for your reply! Now I've solved the previous problem: It did cause by out of memory.  I started a new training on a sentiment analysis dataset, but this time I met a new issue:
['neutral', 'negative', 'positive']
Traceback (most recent call last):
  File "run.py", line 241, in <module>
    main(args)
  File "run.py", line 19, in main
    class_oriented_sent_representations.main(args)
  File "/home/xuanw/HL/MEGClass-main/class_oriented_sent_representations.py", line 175, in main
    temp_repr = np.array([static_word_representations[word_to_index[w]] for w in class_names[cls].split("_")])
  File "/home/xuanw/HL/MEGClass-main/class_oriented_sent_representations.py", line 175, in <listcomp>
    temp_repr = np.array([static_word_representations[word_to_index[w]] for w in class_names[cls].split("_")])
KeyError: 'neutral'

Is that caused by no "neutral" in raw text?

**Comment by pkargupta** [comment]
@ForeverNightmare Yes, we are constructing the vocabulary based on all of the words contained within the corpus (each word is mapped to its respective static representation, which is the average of all its corresponding contextualized representations). This is necessary because we are only provided the class label surface names as extremely weak supervision, so we need some sort of insight into how "neutral" is used within the context of the corpus. If you would like to still use "neutral", you can either manually add it and its corresponding representation to the static representations/vocabulary or choose to replace it with a synonym present within the corpus (e.g. indifferent).
